File: dartxjs.dart
#library('dartxjs.dart');
#import('dart:html');
#import('dart:json');

class DartToJSCommunicator {
  Function _handler;
  
  DartToJSCommunicator() {
    _handler = null;
    window.on.message.add((MessageEvent e) {
      if (_handler == null)
        return;
      try {
        String json_data = e.data;
        Map<String,Object> cmd = JSON.parse(json_data);
        if (!(cmd is Map) || cmd["rcpt"] != "dart") {
          return;
        }
        _handler(cmd["data"]);
      } catch (Exception ex) {
        // ignore data that can't be parsed
      }
    });
  }
  
  void set receiveMessage(Function f) {
    _handler = f;
  }
  
  void sendMessage(Map<String,Object> data) {
    window.postMessage(JSON.stringify({"rcpt":"js", "data":data}), window.location.href);
  }
}

File: dartxjs.js
JSToDartCommunicator = (function() {

	var addEventListener = function(elem, event, f) {
		if (elem.addEventListener) {
			elem.addEventListener(event, f, false);
		} else if (elem.attachEvent) {
			elem.attachEvent('on' + event, f);
		} else {
			throw "couldn't attach event " + event + " to " + elem;
		}
	};

	return {
		sendMessage: function(data) {
			window.postMessage(JSON.stringify({"rcpt":"dart", "data":data}), window.location.href);
		},

		receiveMessage: function(f) {
			addEventListener(window, 'message', function(event) {
				try {
					var pkt = JSON.parse(event.data);
					if (pkt.rcpt == "js") {
						f(pkt["data"]);
					}
				} catch (e) {
					// ignore exceptions
				}
			});
		}
	}
})();
